### ISSUE TITLE
Make att.typography and rend/@rend CSS compatible

### ISSUE THREAD

**Issue by ahankinson** [music-encoding/music-encoding#140]
_From [zupf...@googlemail.com](https://code.google.com/u/113984963995030839842/) on December 20, 2012 06:20:00_

From the description in the specs, I get the impression that the current att.typography set of attributes seems to be modeled after mup's concept of fonts. This is not a concept that is shared by a lot of other applications. Therefore, I suggest to move towards CSS, like we did for color.

The most important issue I currently see is @fontname. While the attribute name makes me think I can use something like "Arial" here, this is already covered by @fontfam. Instead, "Mup-acceptable values: rom, ital, bold, boldital" are suggested. But "ital" is already covered by @fontstyle and "bold" by @fontweight. "rom" is basically equivalent to "normal" (the specs even say 'normal' = 'rom' in Mup), which is also covered by @-fontstyle.

@-rend also has some issues. It makes non-mutually-exclusive features like small caps and underline mutually exclusive.

I'd therefore suggest consequently following CSS, using the properties
- font-family
- font-style
- font-variant
- font-size
- font-weight
- text-decoration

We could continue using the existing attribute names
- fontfam
- fontsize
- fontstyle
- fontweight

and add names for the missing properties, or we could use the CSS property names as attribute names. What we should do in any case is use the same data as CSS.

CSS to my knowledge however does not cover all aspects that @-rend covers, namely:
- circle/box (the CSS border property doesn't seem approprate here because it's not abstract enough)
- sub/sup (requires two properties: font-size and vertical-align)
- dblunderline

This needs to be resolved somehow. (BTW: What is the meaning of the values "quote", "bslash" and "fslash"? Those seem to be names of single characters and no rendering information.)

_Original issue: http://code.google.com/p/music-encoding/issues/detail?id=140_

**Comment by ahankinson** [comment]
_From [kep...@edirom.de](https://code.google.com/u/113787537940489926112/) on December 20, 2012 05:14:28_

I totally agree with Thomas that we should resemble CSS where feasible. Following MUP is certainly not the right track for the standard in general (it might very well be for individual projects). Let's have a look at TEI. The offer almost none of these CSS-like attributes, but instead they have @rend, @style, and @rendition. The last one points to detailed rendering descriptions in the header, which may use CSS or any other language to specify the exact layout rule. The @style does more or less the same, with the only difference that you cannot provide the information whether you use CSS or any other scheme. The only rule on @-rend is that it contains a space-separated list of rendition rules (in other words, each rule must not contain whitespace). 

In MEI, we have some more attributes, in addition to the ones Thomas pointed out. We have @rotation, which would be somewhere in the realm of CSS3, we have @rend with a closed list of values not addressed by @fontfam etc., such as underline, we have @valign (which may in fact interfere with @rend="sub"), @halign(which more or less mimics css:text-align), and we have @-altrend, which takes all kinds of values. 

I'm not sure how to put all this together in a more consistent way. I see the point of deprecating @fontname, although it does not enforce the values Thomas mentioned (they're probably only a leftover from earlier days, when Perry used the descriptions of the elements to remind him of best practices "for his use"). I also still see a benefit in putting the most frequent CSS properties in specific attributes like @fontfam. They don't have to have the exactly same name as in CSS, as long as it stays clear what they mean. I think it's more important that the values _inside_ match CSS, which is not always true right now (I'm especially worried about @-fontsize, which has to be expressed in printer's points, without the option of specifying alternative units). 

The current values in @rend may have the benefit of not being overly specific, as Thomas pointed out for circle and box. They seem to be more valuable to transcribe a source than to specify the layout for a rendition of the file. In their current state, they are helpful for some, but may not be sufficiently detailed for others. I don't think we can resolve that, and maybe it's better to distinguish the use cases better. TEI's explicitly states that @rend should be used to describe existing sources, whereas @-style should be used for displaying the code. Shouldn't we move in that direction as well?

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on December 20, 2012 05:44:52_
1. You're right about fontname -- it's not useful anymore.  As you suggest, we can simply remove it.
2. @-rend makes the recording of typographic features mutually exclusive _intentionally_.  The idea was to force the use of multiple <rend> elements.  For example, text that is bold, circled, and underlined would require markup like this -- 

<rend rend="bold">
  <rend rend="circle">
    <rend rend="dblunderline">text</rend>
  </rend>
</rend>

This was necessary when using a DTD in order to constrain the values of @-rend to a given list.  However, it's no longer necessary, given the capabilities of RNG and XSD schema.  This is why this issue was filed.
1. I'm reluctant to add attributes to <rend> for each typographical property, but that might prove to be the easiest approach.  How about constraining the value of @-rend to a space-separated list of properties?  For example,

<rend rend="fontweight:bold circle underline"> or
<rend rend="fontweight:bold; circle; underline">

Some of these values might be CSS, some might not.  Because we need properties like enclosure, etc. this can't be absolutely CSS-compliant, only "CSS-like" or "CSS-compatible".

(The semicolons don't matter, only the spaces. "fontweight:bold" is different than "fontweight: bold" with or without the semicolon because XML always tokenizes attribute values on spaces.)
1. The value "quote" indicates that the text is wrapped by quote characters.  This has been a topic recently on TEI-L (although in the context of <quote> and <q> elements).  I think we should just drop this value and let the user put quotes inline if they're necessary.

"bslash" and "fslash" are special forms of strikethrough using "\" and "/", respectively.

The values of @-rend are a hodge-podge of HTML, commonly-used TEI, and other values.  It's time we take up the issue, but I don't know exactly how to proceed.  I'm open to all suggestions.

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on December 20, 2012 05:45:38_

I'm happy to move forward, but like the rest of you I'm not certain which way to go.  CSS alone won't help us, since we have additional needs.  The TEI approach might be beneficial, but may introduce more complexity than we want to admit.

**Comment by ahankinson** [comment]
_From [kep...@edirom.de](https://code.google.com/u/113787537940489926112/) on December 20, 2012 06:00:20_

PR: 
> As I said in my earlier message, I'm happy to move forward, but like the rest of you I'm not certain which way to go.  CSS alone won't help us, since we have additional needs.  The TEI approach might be beneficial, but may introduce more complexity than we want to admit.

What if we make a distinction between description of an original and specification of a layout like TEI does in the first place? I think this would be extremely helpful. On the input side, we don't have to be overly specific, and in fact, I think CSS is not helpful here. On the output side, CSS seems like the way to go for almost everything. 

A @style attribute could take any string, as long as it complies to CSS. TEI has an example which reads style="text-align: center; font-variant: small-caps". Actually, this matches exactly with the @-style attribute in HTML, and I see no reason not to use that. 

I'm less sure on what that would mean for the other attributes. When describing a source, I want to specify fontsize using terms like "normal", "small", "smallest" etc, but not 14px or 1.2em. Both styles are actually allowed within CSS. The font may also be described with terms like "latin", "sütterlin", "spaced" etc, which is less covered by CSS. This clearly needs more thinking. 

I'd like to hear more opinions on the usefulness of splitting up between input code and output code. Resolving the details is something I would like to delegate to the Layout SIG. Speaking of which – is someone preparing a submission for a panel for this SIG at the Mainz Conference? It would be terrific to get that. We should make sure that this SIG doesn't get lost on the way, it seems to important. Also, should we set up something for the SIG? I would like to integrate a WIKI on our website anyway, would that be helpful? Separate mailing lists? Coffee, cookies?

Best,
jo

**Comment by ahankinson** [comment]
_From [andrew.hankinson](https://code.google.com/u/andrew.hankinson/) on December 20, 2012 09:30:31_

I agree with Johannes -- if we're going with CSS, we should make sure that we don't confuse the matter by using it in a different way from HTML, so add the @style attribute (This may also mean supporting an @-class attribute for external stylesheets, but that's slightly tangential here.) If it looks like a duck and walks like a duck, we should make sure we make it act like a duck. (quack!)

It would be helpful to have an actual list of our 'additional needs' and how CSS does not cover it. I suspect that CSS3 (and especially the advanced CSS selectors) would cover most of what we would need to do for formatting.

Since source description is essentially a semantic issue (you're "marking up" an existing document) we should definitely make the distinction between document encoding and document presentation. (What I think Johannes calls "input" and "output" code).

**Comment by ahankinson** [comment]
_From [kep...@edirom.de](https://code.google.com/u/113787537940489926112/) on December 20, 2012 10:33:43_

AH:
> It would be helpful to have an actual list of our 'additional needs' and how CSS does not cover it. I suspect that CSS3 (and especially the advanced CSS selectors) would cover most of what we would need to do for formatting.

The problems arise when CSS is actually _too_ specific. I may jsut want to say that something is boxed, but I don't want to bother about the differing border width, color, and radius. Or that the font is spaced, but I don't want to track the exact spacing. So on the document encoding side (which Johannes called "input"), more generic terms than CSS offers may be more suitable. The increasing level of 'specific-ness' in CSS3 and CSS4 doesn't seem to offer that, it actually contradicts. 

AH:
> This may also mean supporting an @-class attribute for external stylesheets
This is what TEI addresses with it's @-rendition system. This attribute just provides a pointer to a style definition in the header, just like HTML classes are picked up by CSS rules. The difference is that the pointer goes in the other direction, but that seems like a technical detail to me.

My current ideal solution (and I'm happy to revise that) would be a semi-controlled list of descriptive terms for document encoding, more or less like we have it in @rend. Semi-controlled means that there should be a list of proposed values, and the flexibility of adding new terms. For document presentation, I'd like to see something CSS-based. This could use @style, and a mechanism like TEI's @-rendition. If there is only document encoding code, it would be up to the processing application to convert that to presentation rules as specified by CSS. The crucial thing is to explain the different scopes of these attributes very clearly. 

Best,
Johannes

**Comment by ahankinson** [comment]
_From [andrew.hankinson](https://code.google.com/u/andrew.hankinson/) on December 20, 2012 11:10:52_

Hm, maybe I was mistaken in my interpretation of your comments. By "input" I thought you meant encoding the original document, i.e., describing a box that's printed around a note. To me, this would go in the markup and not in the style. For "output" I assumed you meant describing only how elements should be displayed in a renderer.

So you wouldn't run into the situation where you have something boxed but don't want to bother with describing it. If you want it to display with a 2px border, you put that in the CSS. If you just want to say that it's boxed, you put it in the markup with no style rules. It's the same as wrapping an element in a generic block-level element in HTML. 

Encoding information about a document would essentially be 'metadata' (like if a printer used a known font face from a certain foundry), I would think that this would be part of the document encoding itself, held in the header. If we then had an electronic replica of that font face we might then choose to have the renderer display the notation in that font face. But I see those as two separate processes.

Similarly, if we were describing coloration we would encode the fact that a note was coloured in the document markup, but would leave it up to a style sheet to actually describe what shade of red we wanted it to render as.

In a 'pure' implementation we would deprecate @-rend all together, since describing how something is rendered would be handled in the CSS and not in the markup. But we would need to figure out first what CSS doesn't have that we need, and second what CSS has that we don't need. I suspect that the list for both will be fairly short, since layout is layout no matter what you're actually laying-out.

I wouldn't think we would even need to control the subset of CSS, since the support for these elements wouldn't necessarily be in MEI but in the renderers themselves. This would end up looking like the tables that show which browsers support which CSS features. So one notation renderer might support the full CSS spec, while another might only support a subset.

**Comment by ahankinson** [comment]
_From [zupf...@googlemail.com](https://code.google.com/u/113984963995030839842/) on December 20, 2012 12:42:33_

I fully support separating description of existing documents and rendering information (and Andrew, I don't think you misunderstood Johannes).

TEI's rendition system seems really straightforward, but I read that this system is intended for encoding the original rendering: http://wiki.tei-c.org/index.php/Formatting_TEI_documents#Original_vs._output_rendering For TEI it's of course much easier to describe the output style using "standard" methods like XLST and plain standalone CSS applied to a TEI document, e.g. using
  <?xml-stylesheet type="text/css" href="xyz.css" ?>
or
  <?xml-stylesheet type="text/xsl" href="xyz.xsl" ?>

We probably _do_ need something more suitable for MEI.  If we're using something along the lines of TEI's rendition system, we should probably change its name so it's not confused with the semantics (i.e. describing the "input"/original/facsimile rather than the output/rendering) it has in TEI. Or we could use the same system for both (which might be convenient), but have different attribute names, like @sourceRendition and @-targetRendition.

JK:
> I'm less sure on what that would mean for the other attributes. When describing a source, I want to specify fontsize using terms like "normal", "small", "smallest" etc, but not 14px or 1.2em.

I think that CSS most of the time offers values that are unspecific enough. In fact, font-size accepts values xx-small, x-small, small, medium, large, x-large, xx-large for "absolute" sizes as well as "larger" and "smaller" for "relative" size changes.  

JK:
> I may jsut want to say that something is boxed, but I don't want to bother about the differing border width, color, and radius.

You don't have to with CSS, just say "border-style:solid".

JK:
>  Both styles are actually allowed within CSS. The font may also be described with terms like "latin", "sütterlin", "spaced" etc, which is less covered by CSS.

When we need new properties or values, I think we should introduce our own "vendor prefix", as suggest by the CSS Working Group: http://wiki.csswg.org/spec/vendor-prefixes#third-party-features This vendor prefix ("-mei-" seems natural) would be applicable for property names and values, this means we could both introduce new properties and extend existing ones in conformance with CSSWG best practices. We could e.g. have a "-mei-border-shape" property that accepts values "box, circle, none" (which, by the way, are the values allowed for @-reh.enclose). Likewise, we could allow "text-decoration:-mei-dblunderline".

I have a bad feeling about adopting @-class and CSS selectors. This looks like really complex stuff. Parsing CSS syntax in a style attribute is really simple--just split at colons and semicolons--but determining CSS matches?  If there are libraries that can help with that, I still feel uneasy because it ties us to them and possibly to certain languages/frameworks/platforms/whatever.

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on December 20, 2012 12:43:54_

Taking the baton from Jo, here's a list of values for @-rend.  They would be used, as Jo described, when a generic description of the source material is needed.

Some values are analogous to CSS:

xx-small | x-small | small | medium | large | x-large | xx-large | smaller | larger | italic | oblique | small-caps | bold | bolder | lighter | underline | overline | line-through | capitalize | uppercase | lowercase | left | right | center | justify | sub | super | ltr | trl 

CSS does _not_ provide these:

dblunderline | dbloverline | bslash | fslash | dblfslash | box | circle

With the exception of box and circle, the terms in the 2nd set represent "embellishments" of values in the 1st set. That is, "dblunderline" and "dbloverline" indicate under/over-lining of some text twice.  "bslash", "fslash", and "dblfslash" are more specific forms of over-strike, as they specify the character to be used and number (in the case of dblfslash). These all have Unicode character equivalents.

"box" and "circle" indicate a container for the text with a particular border -- a square/rectangle in the case of "box" and a circle/ellipse in the case of "circle". No provision is made to describe the enclosing shape's color, size, etc. although I think it's a good assumption that the shape should fit the text, not the other way around. In other words, as the text to be enclosed gets wider, so does the enclosing shape.  Therefore, the value "circle" doesn't mean the text is inside an actual circle, but that the text has been "circled".  Maybe it would be clearer to change these values to "boxed" and "circled".

Any number of these values should be allowed to occur within @rend.  In other words, @-rend should take a list of tokens, not just one.

@-rend does not address "output" description; that is, what something rendered from the MEI file should look like.  For that I think we should remove the existing font-related attributes (like fontfamily, fontsize, etc.) and add a new "style" attribute that allows any CSS.  I don't know of any way to control the value of this new attribute, so the user will be on his own to make sure it's not gibberish.

@-style can also be used where the "input" can be usefully described using CSS, as is the case when the source is in another music description format. The presumption is that the output will simply mirror the input description.

I think we can table the discussion of @rendition (a pointer to a style specification somewhere in the header). That capability is useful when there's a lot of style repetition. But I anticipate that there's so much variability in how text is handled (even in a single piece), that one would end up with very much the same number of descriptions in the header as there are chunks of text in the document.  The effect would be to simply move the style specifications from the individual elements to the header.  But, of course, I could be wrong -- in which case, we could still take up @-rendition later.

Sound good?

--
p.

**Comment by ahankinson** [comment]
_From [zupf...@googlemail.com](https://code.google.com/u/113984963995030839842/) on December 27, 2012 08:28:49_

PR:
> @-style can also be used where the "input" can be usefully described using CSS, as is the case when the source is in another music description format. The presumption is that the output will simply mirror the input description.

Is this something that we should analyze more consequently in the context of the ongoing layout discussion?  I understand that the current idea of describing different rendering layouts is using essentially the same means that are used to describe the source (or the "main" data), and this would be applicable here as well.

This means: The attributes with CSS-like data could be the same for both the source and the layout.  Everything from the source description is mirrored to the layout description, unless overridden.  In the source, we would usually use less specific values (like "small"), although we're free to use very specific values (like "12pt" or "1.1cm").  If CSS is too specific (or not specific enough) or doesn't come with everything we need, we can introduce prefixed properties and values of our own and still be 100% CSS compatible (as I suggested in post `#8`).

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on January 02, 2013 09:05:31_

This issue will undoubtedly play a part in the layout discussion.  However, as I'm currently working on revising the MusicXML -> MEI transformation and I expect the layout discussion to be continued well past the end of the current grant, I'd like to separate this issue from the layout discussion as far as possible now and resolve it as quickly as possible.   

Is everyone (relatively) happy with 
1. removing the existing font-related attributes (like fontfamily, fontsize, etc.) and adding @-style that allows any CSS, and
2. permitting @-rend to take any number of the following values?

xx-small | x-small | small | medium | large | x-large | xx-large | smaller | larger | italic | oblique | small-caps | bold | bolder | lighter | underline | overline | line-through | capitalize | uppercase | lowercase | left | right | center | justify | sub | super | ltr | trl | dblunderline | dbloverline | bslash | fslash | dblfslash | box | circle

This list could also be made open if necessary; that is, values other than those in the list would be acceptable (but not interoperable).

This proposal doesn't directly address the difference between description of the source material and description of the output, but I don't think that's necessary at the moment.  Currently, both @rend and @style describe the input, but adding @-style allows capture of CSS-like description in CSS terms; that is, without a lot of transformation.

If no one has objections, I'll roll this change into the 2013 changes by the end of the week.

Happy New Year to everyone!

**Comment by ahankinson** [comment]
_From [andrew.hankinson](https://code.google.com/u/andrew.hankinson/) on January 02, 2013 09:10:39_

Go for it.

Should we open another issue for a conversation about the larger layout issues?

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on January 02, 2013 09:28:50_

Thanks for the +1.

Please do start another issue for the layout questions.  Separating layout from content is going to be a much larger issue that's going to take a while to resolve.  I hope, however, that in the meantime, the resolution of several smaller/easier problems will whittle it down to manageable size. For example, the resolution of this issue and the page size question will help some.

**Comment by ahankinson** [comment]
_From [zupf...@googlemail.com](https://code.google.com/u/113984963995030839842/) on January 02, 2013 12:15:56_

Does this mean we will have two equivalent options for defining text rendering for most of the listed features? (E.g. rend="medium" and style="font-size:medium")

**Comment by ahankinson** [comment]
_From [kep...@edirom.de](https://code.google.com/u/113787537940489926112/) on January 02, 2013 13:14:12_

+1 for the changes. I would understand them slightly different, though, and maybe this matches Thomas' concerns. For me, @style should describe output, not existing sources. That way, @rend and @style do not interfere. Also, do we plan to keep @altrend? Do we need to open the list of @rend values, when we have it, or will we deprecate that in favor of an open list? I'd vote for keeping the second attribute, as processing @-rend seems much easier when it has a finite number of values to take. By the way, I'd suggest "spaced" or a better term for that. 

On the layout issues: is someone preparing a session proposal for the conference in May? Someone could also ask for a workshop on this? Workshops don't have to be all lecturing, but could also be an opportunity for various groups to gather… Maybe preparing these things is the most pressing thing to do, as there is a deadline approaching ;-)

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on January 02, 2013 13:40:48_

The short answer is "yes".

The longer answer is "yes until we don't."  That is, we will have two equivalent options until everyone can agree on a single one. I don't see that happening in the immediate future as there are those who want to emulate a TEI-like methodology (that is, non-CSS) and there are those who desire a CSS-only (plus enhancements) solution.  Although I've seen the error of my past ways and now prefer the CSS-only approach, I see the logic in the other one as well in that it keeps MEI in step with TEI.

I just had a "Doh!" moment -- How about implementing the CSS-only approach in @rend and putting the values that are currently allowed in @rend in @altrend instead?  Since @altrend is already defined as non-interoperable, the "loosey-goosey" TEI-like values can go there.  There's still the potential for semantic overlap between @rend and @altrend, but that possibility already existed anyway. It always was the developer's choice to support @altrend; that is, syntax other than CSS. But we'll be making a statement that CSS is preferred by choosing its syntax over the current values for @-rend.

Better?

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on January 02, 2013 13:47:24_

Sorry, didn't see Jo's comments.  But it looks like we're coming closer to an agreement. The list of values for @-altrend can be closed by default if that's easier to support.

The association of "output" with "@style" can be side-stepped by not creating @-style in the first place.  :-)

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on January 04, 2013 05:45:14_

Any more comments on this issue, especially on the latest proposal for a "solution"?

**Comment by ahankinson** [comment]
_From [kep...@edirom.de](https://code.google.com/u/113787537940489926112/) on January 04, 2013 06:03:03_

I had the impression that people were happy to get @style, as it matches closely with HTML and TEI. But it's true that TEI uses @-style for transcription ("input") only, not for rendition ("output"). I just had a look, and TEI doesn't seem to offer attributes for rendition at all ( http://www.tei-c.org/release/doc/tei-p5-doc/en/html/ref-att.global.html ). 

For parts of the MEI community it is certainly desirable to store output information within the encoding. To me, it seems like a good idea to clearly separate things in order to avoid any potential confusion. We don't have to mess things up in the first place just to get a reason for revising them later ;-) 

This could be side-stepped by defining @rend as CSS-like input description, @altrend as free-form input description, and @style as CSS-like output description (with the potential of vendor-prefixed additions). This might be the cleanest solution, though it has the side-effect of deprecating existing files, which use @rend in non-CSS-compliant ways. But: is anyone of us really using @-rend significantly? 

Axel, is your use of @rend compliant with CSS, or would you expect to face a lot of problems when having to make it CSS-compliant? In Detmold, we have no significant data using @-rend, so I assume that you're the one with biggest horse in the race…

**Comment by ahankinson** [comment]
_From [zupf...@googlemail.com](https://code.google.com/u/113984963995030839842/) on January 04, 2013 07:14:50_

If we have one attribute for source description and one for output description, I think that should be more clearly reflected in the attribute names, like by "sourceStyle" or "renderStyle". This would avoid misuse and confusion with the TEI style attribute.

**Comment by ahankinson** [comment]
_From [a...@kb.dk](https://code.google.com/u/101822190470135414035/) on January 04, 2013 07:28:50_

We do use @-rend for text formatting, but as the values do translate 1:1 to CSS equivalents, changing them is no big deal, really. And once it is done, it probably would make it even easier for us working with the texts.

/axel

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on January 04, 2013 07:48:51_

I'm getting dizzy with all the input/output description possibilities!

Can we step back for a moment and answer the question "Why should output description be included in the encoding?"

I believe we should keep in mind that, despite a few exceptions, such as @x, @y, @x2, and @y2, MEI _describes the input document_. For example, @place, @color, @-size, etc. describe the source material. But, they are also "hints" for rendering.  As "hints" they may or may not be passed though to some output format.  But, they are not a description of the output format, per se.

I don't believe the MEI encoding is the proper place to store information about the output -- and I wonder if @x, @y, @x2, and @-y2 shouldn't be removed.

I wasn't aware that Axel was using @rend.  If Axel or anyone else is using @rend, then my proposal from yesterday has to be reconsidered in the light of the ever-present bug-a-boo "compatibility".  We might be better off to adopt the TEI approach -- @rend for non-CSS description, @style for description in CSS syntax, and drop @altrend to avoid confusion.  Or just use @rend and @altrend but flip their meanings -- @rend is for CSS, @altrend is for non-CSS.  Any prior use of @rend would go into @-altrend.

Returning to the philosophical question of "What are we representing?", it seems to me that even MerMEId is not describing a physical output document, but an abstract input.  It's what the input document would look like if it actually existed.  This is not unlike using MEI to describe a born-digital piece of notation where no physical document actually exists.  At first blush it appears that the markup describes the output document that might eventually be created, but I don't believe this is actually the case.

**Comment by ahankinson** [comment]
_From [andrew.hankinson](https://code.google.com/u/andrew.hankinson/) on January 04, 2013 07:56:12_

> I don't believe the MEI encoding is the proper place to store information about the output -- and I wonder if 
> @x, @y, @x2, and @-y2 shouldn't be removed.

I think I'll need a bit of clarification on this comment. Where would the proper place to store information about the output be?

**Comment by ahankinson** [comment]
_From [kep...@edirom.de](https://code.google.com/u/113787537940489926112/) on January 04, 2013 08:09:54_

< I think I'll need a bit of clarification on this comment. Where would the proper place to store information about the output be?

See, that's the reason why we urgently need this Layout SIG. I'm sure that Laurent would join you in your criticism of Perry's statement. I'm feeling uncomfortable with the current situation as well. 

My idea was to provide a way of encoding output information in the MEI core. This could then be taken up by whatever the layout tree comes up with (layout trees, page-based approach, cross-references, separate files, whatever). No one would have to use @-style (as meant for output), but it would give us the possibility of serializing everything back into plain MEI files (files which explicitly describe exactly one specific rendition). I hoped that we could implement this as a starting point for the more complex layout decisions, but obviously we need to go back to step 1. 

Perry, do I get you right when you say we should remove all explicit output-related info by defining everything as a source, even the outputs of a rendition process? Isn't that the consequence of what you say?

(sorry for repeating myself: Is someone preparing a submission to request a session or workshop on this? Perry and I can't do that, and it seems worthwhile to bang our heads together…)

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on January 04, 2013 08:12:24_

Sorry for going on about this, but if it wasn't clear from my earlier message, I'm strongly not in favor of adding output description to MEI -- that should be addressed elsewhere, perhaps as someone suggested earlier (Craig?) in a "CSS for music".

So, given Axel's flexibility and in the interest of a prompt solution, let's go with the following, shall we? 
- _no new @sourceStyle and @-renderStyle attributes_
- retain @-rend for non-CSS description of the _input document_
- create new @-style for description of the _input document_ in CSS (or other formalized language) syntax

This addresses the initial problem Thomas identified (and that I was having trouble with in my XSLT) now, while also maintaining compatibility with TEI and without entangling it with the issue of output document description.

**Comment by ahankinson** [comment]
_From [andrew.hankinson](https://code.google.com/u/andrew.hankinson/) on January 04, 2013 08:24:53_

CSS is an output formatting specification. The way I see it, anything that's in the original document (the "Input") belongs in the markup. Anything that is used to render output belongs in the CSS. Adding @-style to describe input seems to mess up this model.

If we're explicitly _not_ going to deal with output, in the interests of future work with the layout group I would propose we simply reserve @style (but not use it) for the eventual output, and allow CSS attributes on @-rend (as proposed) to describe the document layout. This breaks compatibility with TEI, but I would rather keep the CSS-for-layout paradigm than maintain compatibility with TEI.

As for proposing a session on the Layout SIG, I haven't heard anything about it. I would be willing to put together the abstracts and submit the proposal, but I would need people to get in touch with me (Laurent? Thomas? Axel? Maja?) to let me know they're writing an abstract for it.

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on January 04, 2013 08:28:50_

>> I don't believe the MEI encoding is the proper place to store information about the output -- and I wonder if 
>> @x, @y, @x2, and @-y2 shouldn't be removed.

> I think I'll need a bit of clarification on this comment. Where would the proper place to store information about the output be?

I can't say exactly, but not in the MEI file itself.  CSS exists for the separation of content and presentation for text.  Something similar for music would be great.  But, peppering MEI with such things is not a valid approach, in my opinion.  That's a step in the wrong direction.

By "not in the MEI file" I don't mean that there can't be a place to include the musicCSS, similar to the way a CSS style sheet can be embedded in an HTML document.  That actually makes it clear that the CSS is describing the output.  What we're doing in @rend and @-style is clearly something different; that is, using CSS to describe the _input_ document.

I'm not "on a tear" to remove @x, @-y, etc.  I simply used these examples of where MEI admits the description of output from it.  We need not remove them if they're currently useful, but we should be wary of going further down that path.

Clearer?

**Comment by ahankinson** [comment]
_From [andrew.hankinson](https://code.google.com/u/andrew.hankinson/) on January 04, 2013 08:37:15_

I think I mostly agree with you, with the exception of using @-style. If, down the road we bring in our "music CSS" system, the most appropriate place to store element-level output styles would be in that attribute. 

I think we can avoid a lot of compatibility issues down the road by simply reserving this attribute for later description of the output, and making @-rend the only option for describing input styles.

(BTW: what does "rend" stand for?)

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on January 04, 2013 08:38:08_

> CSS is an output formatting specification. The way I see it, anything that's in the original document (the "Input") belongs in the markup. Anything that is used to render output belongs in the CSS. Adding @-style to describe input seems to mess up this model.

Our discussion here mirrors (as I remember it) the discussion on TEI-L regarding the addition of @style.  It went something like -- "What?!  Use @-style to capture data about the input?  That's not how HTML uses it. That's just crazy talk!"  But that's what TEI decided to do.  Now we have to decide whether to emulate their decision.  I'm not arguing for following their choice, but simply stating that we either have to follow it or not -- we can't do both.

Along with Andrew, I'm fine with not following TEI.  I just want to "separate concerns" by putting output description off the table for the time being.

**Comment by ahankinson** [comment]
_From [kep...@edirom.de](https://code.google.com/u/113787537940489926112/) on January 04, 2013 08:38:25_

AH
> As for proposing a session on the Layout SIG, I haven't heard anything about it. I would be willing to put together the abstracts and submit the proposal, but I would need people to get in touch with me (Laurent? Thomas? Axel? Maja?) to let me know they're writing an abstract for it.

Maybe it's better to have a workshop on this. Workshops don't have to be tutorials, but could also be working group meetings. Benefit: We don't need a very formal contribution right now, just something for the records. They don't go through the double-blind reviewing process. It doubt we're able to fill an official session (which would be considered in the proceedings as well), although there might be some valuable materials. I'd suggest to have all such contributions with specific proposals submitted as individual papers. I'm sure we can group them into a session later on if feasible. Sounds reasonable?

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on January 04, 2013 08:43:02_

>(BTW: what does "rend" stand for?)

From the TEI Guidelines:

"@rend (rendition) indicates how the element in question was rendered or presented in the source text."

So, it doesn't mean "take the value in this attribute and use it to render some output", although it doesn't preclude that if your goal is to emulate the visual presentation of the original.

**Comment by ahankinson** [comment]
_From [kep...@edirom.de](https://code.google.com/u/113787537940489926112/) on January 04, 2013 08:43:15_

AH:
> BTW: what does "rend" stand for?

rend stands for compatibility with TEI… seriously, the translation into rendition or something similar is misleading in our current discussion, and we really shouldn't touch that oddity…

To me, it seems like allowing arbitrary values in @altrend and recommending CSS (or maybe other defined syntax) in @rend the safest way to go. After the whole discussion, I'd also vote for side-stepping the issue by not implementing @-style in the first place (like Perry suggested yesterday). We should only care about the input side right now, without setting up additional hurdles for future developments…

**Comment by ahankinson** [comment]
_From [andrew.hankinson](https://code.google.com/u/andrew.hankinson/) on January 04, 2013 08:47:58_

RE: Layout workshop. I can send an e-mail to MEI-L asking if anyone is interested in participating. Is this ok?

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on January 04, 2013 09:03:36_

> We  should only care about the input side right now, without setting up  
additional hurdles for future developments…

It was a long way to go, but I'm glad to get there!  :)

Now, I wish there was a way to mandate CSS in @rend. There's no reason to attempt to parse the CSS in @-rend, but can a regex be used to determine if there's actually CSS there and not something else?

**Comment by ahankinson** [comment]
_From [kep...@edirom.de](https://code.google.com/u/113787537940489926112/) on January 04, 2013 09:07:52_

Not sure if we're already there. Others might still complain… I wouldn't try to write a regex for CSS in general. You remember the issues caused by the regex which should od nothing else but color validation? I think a proper statement in the guidelines has to be enough.

**Comment by ahankinson** [comment]
_From [zupf...@googlemail.com](https://code.google.com/u/113984963995030839842/) on January 04, 2013 09:42:39_

+1 for disregarding output rendering right now.

If we want to simplify validation and processing, I'm still thinking that using separate attributes for different "properties" (basically like we do it right now) would not be the worst thing. Quoting from post `#0`:

> We could continue using the existing attribute names
>  - fontfam
>  - fontsize
>  - fontstyle
>  - fontweight
>
> and add names for the missing properties, or we could use the CSS property names as attribute names.

This means, I think we could go for one of the following options:
1. Use the existing attribute names:
      fontsize="smaller" fontstyle="italic"
2. Use the CSS property names as attributes (this is how SVG does it):
      font-size="smaller" font-style="italic"
3. Use @rend like HTML @-style (but describing the input):
      rend="font-size:smaller; font-style:italic"

I have sympathies for `#2`, although this introduces some inconsistencies with the hyphenless naming we usually have, but it simplifies mapping to the CSS properties that are addressed. `#1` and `#2` both allow specifying exactly what properties and values are supported. Like this, we don't lead people into temptation to misuse this feature, possibly throwing all of the 160 or so CSS properties that HTML knows of in there (not to speak of the additional ones that are introduced by SVG and maybe other standards).

**Comment by ahankinson** [comment]
_From [andrew.hankinson](https://code.google.com/u/andrew.hankinson/) on January 04, 2013 10:40:35_

If I were to vote on Thomas' list, I would prefer `#3`.

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on January 04, 2013 11:04:34_

Andrew said:
> If I were to vote on Thomas' list, I would prefer `#3`. 

Thomas said:
> `#1` and `#2` both allow specifying exactly what properties and values are supported. Like this, we don't lead people into temptation to misuse this feature, possibly throwing all of the 160 or so CSS properties that HTML knows of in there (not to speak of the additional ones that are introduced by SVG and maybe other standards).

Yes, `#1` and `#2` are explicit, but as noted above, but even if we don't support all ~160 CSS properties, options 1 and 2 can eventually lead to a tremendously large number of attributes. So, I'm inclined toward `#3`.

I'm also not sure that it's misleading to allow all CSS in @rend. @-rend should allow as full a description as is necessary.  Again, that's separate from what can / should be rendered in the output.

Johannes said:
> Not sure if we're already there. Others might still complain… 

I think we've reached a good resting point. :)

> I wouldn't try to write a regex for CSS in general. You remember the issues caused by the regex which should [do] nothing else but color validation? I think a proper statement in the guidelines has to be enough.

I wasn't suggesting a regex that validates the CSS, as with color, but one that simply verifies that the value is likely CSS -- something along the lines of checking that the value in @-rend a) consists of semi-colon separated tokens, and b) each token takes the form of a colon separated name and value pair.  But I'm happy to leave well enough alone and go with a recommendation in the documentation.

**Comment by ahankinson** [comment]
_From [zupf...@googlemail.com](https://code.google.com/u/113984963995030839842/) on January 04, 2013 12:11:50_

I'm totally fine with my option `#3`, as this seems to be the solution favored by most. In the documentation, should we list the properties that we regard as supported?  (For the values, we could just point to the CSS specs.)

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on January 04, 2013 13:53:16_

I believe giving a list of properties "we regard as supported" is unnecessarily limiting.  It also opens up another can o' worms, since we've yet to say what our position is with regard to CSS vendor prefixes vs. ad hoc values.

Now that we've agreed to use @rend for CSS and @-altrend for non-CSS, what do we mean exactly by "non-CSS"?  In fact, by providing vendor prefixes, isn't CSS already capable of handling properties that are currently non-CSS compliant (in other words that are not interoperable)?

What _I_ mean by "non-CSS" is data expressed in a form other than name:value pairs; that is, as space-separated values.  As I said earlier, these values may have semantic overlap with those in @-rend.  I don't think we can achieve semantic separation here, only separation of the syntax in which the values are expressed.

I think we could provide a non-exhaustive list of CSS properties as examples of CSS syntax that could go in @rend.  We could also outfit @altrend with an open value list that contains the values I suggested in post `#9` for @-rend but that also accommodates ad hoc values such as "foo", where "foo" means something to the encoder (and his custom-built renderer).

I know this sends us right back into the tail-spin we've experienced over the last couple of days, but we can't actually expect users to create their own CSS vendor prefixes if we eliminate @altrend, can we?  That doesn't sound realistic to me.  But if we provide @altrend, we must accept that there will be functional overlap between it and @-rend.

If anyone can suggest a good way out of this conundrum, I'm all ears.

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on January 11, 2013 08:33:43_

I've begun trying to implement something that matches our resolutions here, so here's an update --

I've used the original purpose of @rend vs @altrend; that is, formalized vs. ad hoc description, to guide my thinking.  Putting CSS in @rend and using @-altrend for the MEI-defined values plus any user-defined values blurs this separation.

Creating yet another attribute for ad hoc values didn't seem like a very good path, so I accommodate both CSS and MEI-defined values in @-rend.  So, either of the following is acceptable --

<rend rend="large center top">Wie Melodien zieht es mir</rend>

<rend rend="font-size: 24pt; text-align:center; vertical-align: top;">Wie Melodien zieht es mir</rend>

One cannot, however, intermingle CSS and MEI syntax --

<rend rend="large text-align:center; top">Wie Melodien zieht es mir</rend>

Of course, either set of values may be easily removed by deleting data.CSS or data.TEXTRENDITION in an ODD customization.

I'm also just realizing that accommodation of CSS in @rend of <rend>; that is, for inline text, causes a "ripple effect" to other elements that need visual description.  For example, should CSS also be allowed on <line> or other visual elements that aren't text?  How far should this war on visual attributes go?  I don't think these are issues that we can take up now, so we'll just have to acknowledge that this change will a great deal of inconsistency (use of CSS to describe text, but individual attributes like @color to describe other symbols).  Sorry to pick at this wound some more, but perhaps leaving @-rend alone right now and focusing on making sure the attributes in att.typography are in order would be a good fallback position.

I've attached the ODD modification file I've been using to test this and other upcoming changes, as well as an RNG schema generated from it.  Try it and let me know what you think.

**Attachment:** [mei2013Customization.xml mei2013Customization.rng](http://code.google.com/p/music-encoding/issues/detail?id=140#hc42)

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on March 14, 2013 12:17:51_

This issue was closed by revision r864 .

**Status:** Fixed

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on March 14, 2013 13:54:17_

This issue was closed by revision r867 .

**Comment by ahankinson** [comment]
_From [pd...@virginia.edu](https://code.google.com/u/103686026181985548448/) on March 14, 2013 14:14:12_

This issue was closed by revision r869 .
